swayfmt/src/utils/map/newline.rs

```diff
@@ -149,18 +143,65 @@ fn add_newlines(
         .skip(1)
         .zip(formatted_newline_spans.iter().skip(1))
     {
-        if let Some(newline_sequence) = get_newline_sequence_between_spans(
-            previous_unformatted_newline_span,
-            unformatted_newline_span,
-            &newline_map,
-        ) {
-            offset += insert_after_span(
-                previous_formatted_newline_span,
-                newline_sequence,
-                offset,
-                formatted_code,
-                newline_threshold,
-            )?;
+        if previous_unformatted_newline_span.end < unformatted_newline_span.start {
+            let snippet = &unformatted_code
+                [previous_unformatted_newline_span.end..unformatted_newline_span.start];
+
+            let start = previous_unformatted_newline_span.end;
+
+            // Here, we will try to insert newlines that occur before and after comments.
+            // The reason we do this is because comments aren't a part of the AST, and so they aren't
+            // collected as leaf spans - they simply exist between whitespaces.
+            if let Some(start_first_comment) = snippet.find("//") {
+                // Insert newlines that occur before the first comment here
+                if let Some(newline_sequence) = first_newline_sequence_in_span(
+                    &ByteSpan {
+                        start,
+                        end: start + start_first_comment,
+                    },
+                    &newline_map,
+                ) {
+                    let at = previous_formatted_newline_span.end + offset;
+                    offset +=
+                        insert_after_span(at, newline_sequence, formatted_code, newline_threshold)?;
+                }
+
+                // Insert newlines that occur after the last comment here
+                if let Some(start_last_comment) = snippet.rfind("//") {
```

Comment:
Do we need to handle `///` as well here? Otherwise the third `/` would be considered the start of the comment.

Reply:
`///` are considered a valid leaf span in the AST, so it won't be picked up here since we're searching between leaf spans.